File: java/WordLadderII.java
/*
 * Filename: WordLadder.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 29 Jun 2015 03:31:57 PM EDT
 */

public class Solution {
	public List<List<String>> findLadders(String start, String end, Set<String> dict) {
		dict.add(end);
		dict.remove(start);

		List<List<String>> res = new ArrayList<List<String>>();
		HashMap<String, Set<String>> parents = new HashMap<String, Set<String>>();

		Queue<String> queue = new LinkedList<String>();
		Queue<String> queue2 = null;
		queue.add(start);

		while (!queue.isEmpty()) {
			queue2 = new LinkedList<String>();

			for (String s: queue) {
				if (s.equals(end)) {
					Stack<String> path = new Stack<String>();
					path.add(end);
					getPath(start, end, path, parents, res);
					return res;
				}

				for (String t: getNbr(s, dict)) {
					queue2.add(t);

					if (parents.containsKey(t)) {
						parents.get(t).add(s);
					} else {
						Set<String> prts = new HashSet<String>();
						prts.add(s);
						parents.put(t, prts);
					}
				}
			}

			dict.removeAll(queue2);
			queue = queue2;
		}

		return res;
	}

	private void getPath(String s, String t, Stack<String> path, HashMap<String, Set<String>> parents, List<List<String>> res) {
		if (s.equals(t)) {
			List<String> rev = new ArrayList<String>();
			rev.addAll(path);
			Collections.reverse(rev);
			res.add(rev);
			return;
		}

		for (String prt: parents.get(t)) {
			path.push(prt);
			getPath(s, prt, path, parents, res);
			path.pop();
		}
	}

	private Set<String> getNbr(String s, Set<String> dict) {
		Set<String> nbrs = new HashSet<String>();
		char[] arr = s.toCharArray();

		for (int i = 0; i < s.length(); i++) {
			char old = arr[i];

			for (char c = 'a'; c <= 'z'; c++) {
				if (arr[i] == c) {
					continue;
				}

				arr[i] = c;
				String t = new String(arr);

				if (dict.contains(t)) {
					nbrs.add(t);
				}
			}

			arr[i] = old;
		}

		return nbrs;
	}
}

File: cpp/word_ladder_ii.cc
/*
 * Filename: word_ladder_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 07 Aug 2015 06:31:40 PM EDT
 */

#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <queue>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<string> > findLadders(
				string start,
				string end,
				unordered_set<string>& dict)
		{
			unordered_set<string> level;
			unordered_set<string> level2;

			mapping.clear();
			res.clear();
			path.clear();

			dict.insert(end);
			path.push_back(end);
			level.insert(start);

			while (!level.empty())
			{
				for (auto it = level.begin(); it != level.end(); it++)
				{
					dict.erase(*it);
				}

				for (auto it = level.begin(); it != level.end(); it++)
				{
					get_nbr(*it, level2, dict);
				}

				if (level2.find(end) != level2.end())
				{
					output(start, end);
					return res;
				}

				level.clear();
				level = level2;
				level2.clear();
			}

			return res;
		}

	private:
		map<string, vector<string> > mapping;
		vector<vector<string> > res;
		vector<string> path;

		void get_nbr(
				string s,
				unordered_set<string>& level2,
				unordered_set<string>& dict)
		{
			for (int i = 0; i < s.length(); i++)
			{
				string s2 = s;

				for (char j = 'a'; j <= 'z'; j++)
				{
					s2[i] = j;

					if (dict.find(s2) != dict.end())
					{
						level2.insert(s2);
						mapping[s2].push_back(s);
					}
				}
			}
		}

		void output(
				string start,
				string end)
		{
			if (start == end)
			{
				reverse(path.begin(), path.end());
				res.push_back(path);
				reverse(path.begin(), path.end());
				return;
			}

			vector<string> v = mapping[end];
			for (auto it = v.begin(); it != v.end(); it++)
			{
				path.push_back(*it);
				output(start, *it);
				path.pop_back();
			}
		}
};
